### PR TITLE
Add cleaner class for third party plugins compatibility 

### PR DESCRIPTION
**Pull request by auban**

### CODE REVIEW

**Comment by svenba** [comment]
@auban please revert the spaces to tabs changes and resolve the conflicts. After that we will review the pull request. Thanks.

**Comment by auban** [comment]
The conflict is resolved. Thanks

**Comment by dramaturg** [comment]
@auban we started go go through all the changes you made. Thank you for all the work you have put into this so far!

Some issues still need to be resolved though:
- We don't think the Autoloader is worth it though so would like to move those classes over to `inc/` and load them just like the other ones.
- You changed a bunch of things in the settings page besides just formatting that will break translations. This should be reverted and just your new settings inserted in between.
- Afterwards, please merge (squash) all your changes into a single commit

File: inc/cache_enabler_disk.class.php [hunk 1]
```diff
@@ -200,7 +184,7 @@ public static function get_asset() {
         }
 
         // check modified since with cached file and return 304 if no difference
-        if ( $http_if_modified_since && ( strtotime( $http_if_modified_since ) >= filemtime( self::_file_html() ) ) ) {
```
Comment (svenba) [inline inc/cache_enabler_disk.class.php]: Please remove this change.

File: inc/cache_enabler_wpml.class.php [hunk 2]
```diff
@@ -0,0 +1,102 @@
+<?php
+
+
+// exit
+defined('ABSPATH') OR exit;
+
+
+/**
+ * Cache_Enabler Woocommerce
```
Comment (svenba) [inline inc/cache_enabler_wpml.class.php]: Odd comment.

File: inc/cache_enabler_woocommerce.class.php [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+        $this->product_set_stock_status($product->get_id());
+    }
+
+    public function product_set_stock_status($product_id) {
+
+        Cache_Enabler::clear_page_cache_by_post_id($product_id);
+    }
+
+    /**
+     *
```
Comment (svenba) [inline inc/cache_enabler_woocommerce.class.php]: Please use clean function comments everywhere.

File: inc/cache_enabler_woocommerce.class.php [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+/**
+ * Cache_Enabler Woocommerce
+ *
+ * @since 1.3.3
+ */
+
+final class Cache_Enabler_Woocommerce {
+
+
```
Comment (svenba) [inline inc/cache_enabler_woocommerce.class.php]: Remove extra newline.

File: inc/cache_enabler_cleaner.class.php [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+        $this->clean();
+    }
+
+
+    /**
+     * Clean process
+     */
+    public function clean() {
+
+
```
Comment (svenba) [inline inc/cache_enabler_cleaner.class.php]: Remove extra newline.

File: inc/cache_enabler_cleaner.class.php [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+        return self::$_instance;
+    }
+
+
+    /**
+     * Cloning is forbidden.
+     *
+     */
+    public function __clone() {
```
Comment (svenba) [inline inc/cache_enabler_cleaner.class.php]: Please remove the __clone and __wakeup functions.

File: inc/cache_enabler.class.php [hunk 7]
```diff
@@ -2118,30 +2160,23 @@ public static function settings_page() {
                             <fieldset>
                                 <label for="cache_excl_ids">
                                     <input type="text" name="cache-enabler[excl_ids]" id="cache_excl_ids" value="<?php echo esc_attr($options['excl_ids']) ?>" />
-                                    <p class="description">
-                                        <?php echo sprintf(__("Post or Pages IDs separated by a %s that should not be cached.", "cache-enabler"), "<code>,</code>"); ?>
-                                    </p>
+                                    <p class="description"><?php _e("Post or Pages IDs separated by a <code>,</code> that should not be cached.", "cache-enabler"); ?></p>
```
Comment (svenba) [inline inc/cache_enabler.class.php]: Please revert this change.

File: inc/cache_enabler.class.php [hunk 8]
```diff
@@ -2118,30 +2160,23 @@ public static function settings_page() {
...
 
                                 <br />
 
                                 <label for="cache_excl_regexp">
                                     <input type="text" name="cache-enabler[excl_regexp]" id="cache_excl_regexp" value="<?php echo esc_attr($options['excl_regexp']) ?>" />
-                                    <p class="description">
-                                        <?php _e("Regexp matching page paths that should not be cached.", "cache-enabler"); ?><br>
-                                        <?php _e("Example:", "cache-enabler"); ?> <code>/(^\/$|\/robot\/$|^\/2018\/.*\/test\/)/</code>
-                                    </p>
+                                    <p class="description"><?php _e("Regexp matching page paths that should not be cached. e.g. <code>/(^\/$|\/robot\/$|^\/2018\/.*\/test\/)/</code>", "cache-enabler"); ?></p>
```
Comment (svenba) [inline inc/cache_enabler.class.php]: Please revert this change.

File: inc/cache_enabler.class.php [hunk 9]
```diff
@@ -2118,30 +2160,23 @@ public static function settings_page() {
...
                                 <br />
 
                                 <label for="cache_excl_cookies">
                                     <input type="text" name="cache-enabler[excl_cookies]" id="cache_excl_cookies" value="<?php echo esc_attr($options['excl_cookies']) ?>" />
-                                    <p class="description">
-                                        <?php _e("Regexp matching cookies that should cause the cache to be bypassed.", "cache-enabler"); ?><br>
-                                        <?php _e("Example:", "cache-enabler"); ?> <code>/^(wp-postpass|wordpress_logged_in|comment_author|(woocommerce_items_in_cart|wp_woocommerce_session)_?)/</code><br>
-                                        <?php _e("Default if unset:", "cache-enabler"); ?> <code>/^(wp-postpass|wordpress_logged_in|comment_author)_/</code>
-                                    </p>
+                                    <p class="description"><?php _e("Regexp matching cookies that should cause the cache to be bypassed. <br>
```
Comment (svenba) [inline inc/cache_enabler.class.php]: Please revert this change.

File: inc/cache_enabler.class.php [hunk 10]
```diff
@@ -1735,7 +1738,7 @@ public static function add_clear_dropdown() {
     public static function add_admin_resources($hook) {
 
         // hook check
-        if ( $hook !== 'index.php' && $hook !== 'post.php' ) {
+        if ( $hook !== 'index.php' AND $hook !== 'post.php' ) {
```
Comment (svenba) [inline inc/cache_enabler.class.php]: Please revert that change.

**Comment by dramaturg** [comment]
@auban, trying to limit the cache cleared to an edited post results in the full cache being cleared. This is regardless of the option "Clear the home page cache on post edit" being ticked or whether there are any additional post IDs being specified. It also looks like the home page cache clearing is always being triggered.

The first issue is probably because you removed `clear_home*` related functionality but the new implementation doesn't work correctly. and proceed to recursively clear the cache in any situation.

Could you check that out?

**Comment by auban** [comment]
Hello,

Your repo is not correctly merged.

I have change the home_url() to home_url('/').

home_url() = not end trailling slash so my new implementation of _clear_dir do a clean directory
home_url('/') = with end trailling slash so clean only file in directory. 

Check your repo diff with this merge request.

Regards,
Auban

**Comment by dramaturg** [comment]
@auban, I've re-tested your code from earlier today and still see the full cache being deleted on post edit - again regardless of the exact settings used.
(In both instances I tested the master branch from RedCastor/cache-enabler)
